File: backend/pkg/database/models/source.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/pgvector/pgvector-go"
	"gorm.io/datatypes"
)

// Source is the GORM model for article sources
type Source struct {
	ID         uuid.UUID       `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	ArticleID  uuid.UUID       `json:"article_id" gorm:"type:uuid;not null;index"`
	Title      string          `json:"title"`
	Content    string          `json:"content" gorm:"type:text;not null"`
	URL        string          `json:"url"`
	SourceType string          `json:"source_type" gorm:"default:web"`
	Embedding  pgvector.Vector `json:"embedding" gorm:"type:vector(1536)"`
	MetaData   datatypes.JSON  `json:"meta_data" gorm:"type:jsonb;default:'{}'"`
	CreatedAt  time.Time       `json:"created_at" gorm:"autoCreateTime"`
}

func (Source) TableName() string {
	return "article_source"
}

File: backend/src/database/models/source.rs
use chrono::{DateTime, Utc};
use diesel::{AsChangeset, Identifiable, Insertable, Queryable, Selectable};
use pgvector::Vector;
use serde_json::Value;
use uuid::Uuid;

use crate::schema::article_source;

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = article_source)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct SourceRow {
    pub id: Uuid,
    pub article_id: Uuid,
    pub title: Option<String>,
    pub content: String,
    pub url: Option<String>,
    pub source_type: Option<String>,
    pub embedding: Option<Vector>,
    pub meta_data: Option<Value>,
    pub created_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Insertable)]
#[diesel(table_name = article_source)]
pub struct NewSourceRow {
    pub id: Uuid,
    pub article_id: Uuid,
    pub title: Option<String>,
    pub content: String,
    pub url: Option<String>,
    pub source_type: Option<String>,
    pub embedding: Option<Vector>,
    pub meta_data: Value,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, AsChangeset)]
#[diesel(table_name = article_source)]
#[diesel(treat_none_as_null = true)]
pub struct SourceChangeset {
    pub article_id: Uuid,
    pub title: Option<String>,
    pub content: String,
    pub url: Option<String>,
    pub source_type: Option<String>,
    pub embedding: Option<Vector>,
    pub meta_data: Option<Value>,
}
